### PR TITLE
[metasrv] Fix flaky test #6242

### LINKED ISSUE #6242: bug: Flaky test about grpc::metasrv_grpc_api::test_auto_sync_addr
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

main branch

### What's Wrong?

Test in PR https://github.com/datafuselabs/databend/pull/6239 failed for

https://github.com/datafuselabs/databend/runs/7064970576?check_suite_focus=true

```shell
---- grpc::metasrv_grpc_api::test_auto_sync_addr stdout ----
Error: timeout after 30s when a leader is observed latest: Metrics{id:1,Follower, term:2, last_log:Some(10), last_applied:Some(LogId { term: 2, index: 10 }), leader:Some(2), membership:{log_id:1-8 membership:[{0, 1, 2}]}, snapshot:None, replication:
Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.58/src/error.rs:533:25
   1: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/result.rs:2105:27
   2: it::grpc::metasrv_grpc_api::test_auto_sync_addr::{{closure}}
             at ./tests/it/grpc/metasrv_grpc_api.rs:376:23
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/future/mod.rs:91:19
   4: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.35/src/instrument.rs:272:9
   5: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/park/thread.rs:263:54
   6: tokio::coop::with_budget::{{closure}}
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:102:9
   7: std::thread::local::LocalKey<T>::try_with
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/thread/local.rs:445:16
   8: std::thread::local::LocalKey<T>::with
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/thread/local.rs:421:9
   9: tokio::coop::with_budget
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:95:5
  10: tokio::coop::budget
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:72:5
  11: tokio::park::thread::CachedParkThread::block_on
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/park/thread.rs:263:31
  12: tokio::runtime::enter::Enter::block_on
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/enter.rs:151:13
  13: tokio::runtime::thread_pool::ThreadPool::block_on
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/mod.rs:90:9
  14: tokio::runtime::Runtime::block_on
             at /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/mod.rs:482:43
  15: it::grpc::metasrv_grpc_api::test_auto_sync_addr
             at ./tests/it/grpc/metasrv_grpc_api.rs:436:7
  16: it::grpc::metasrv_grpc_api::test_auto_sync_addr::{{closure}}
             at ./tests/it/grpc/metasrv_grpc_api.rs:279:7
  17: core::ops::function::FnOnce::call_once
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/ops/function.rs:248:5
  18: core::ops::function::FnOnce::call_once
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/ops/function.rs:248:5
  19: test::__rust_begin_short_backtrace
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/test/src/lib.rs:573:5
  20: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/alloc/src/boxed.rs:1872:9
  21: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/panic/unwind_safe.rs:271:9
  22: std::panicking::try::do_call
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:492:40
  23: std::panicking::try
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:456:19
  24: std::panic::catch_unwind
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panic.rs:137:14
  25: test::run_test_in_process
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/test/src/lib.rs:596:18
  26: test::run_test::run_test_inner::{{closure}}
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/test/src/lib.rs:490:39
  27: test::run_test::run_test_inner::{{closure}}
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/test/src/lib.rs:517:37
  28: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys_common/backtrace.rs:122:18
  29: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/thread/mod.rs:501:17
  30: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/panic/unwind_safe.rs:271:9
  31: std::panicking::try::do_call
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:492:40
  32: std::panicking::try
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panicking.rs:456:19
  33: std::panic::catch_unwind
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/panic.rs:137:14
  34: std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/thread/mod.rs:500:30
  35: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/core/src/ops/function.rs:248:5
  36: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/alloc/src/boxed.rs:1872:9
  37: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/alloc/src/boxed.rs:1872:9
  38: std::sys::unix::thread::Thread::new::thread_start
             at /rustc/cd282d7f75da9080fda0f1740a729516e7fbec68/library/std/src/sys/unix/thread.rs:108:17
  39: start_thread
  40: clone
thread 'grpc::metasrv_grpc_api::test_auto_sync_addr' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', 
```

### How to Reproduce?

No ideas.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Fix flaky test #6242 

## Changelog

- Bug Fix

## Related Issues

Fixes #6242

### CODE REVIEW

File: metasrv/tests/it/grpc/metasrv_grpc_api.rs [hunk 1]
```diff
@@ -410,6 +395,22 @@ async fn test_auto_sync_addr() -> anyhow::Result<()> {
...
 
+        let g = tc1.grpc_srv.as_ref().unwrap();
+        let meta_node = g.get_meta_node();
+        let old_term = meta_node.raft.metrics().borrow().current_term;
+
+        let metrics = meta_node
+            .raft
+            .wait(Some(Duration::from_millis(30_000)))
+            .metrics(
+                |m| m.current_leader.is_some() && m.current_term >= old_term,
```
Comment (drmingdrmer) [inline metasrv/tests/it/grpc/metasrv_grpc_api.rs:406]: `m.current_term >= old_term` will always be `true` IMHO.

**Review by drmingdrmer — APPROVED** [review]

File: metasrv/tests/it/grpc/metasrv_grpc_api.rs [hunk 2]
```diff
@@ -412,6 +397,17 @@ async fn test_auto_sync_addr() -> anyhow::Result<()> {
 
         start_metasrv_with_context(&mut tc3).await?;
 
+        let g = tc3.grpc_srv.as_ref().unwrap();
+        let meta_node = g.get_meta_node();
+
+        let metrics = meta_node
+            .raft
+            .wait(Some(Duration::from_millis(30_000)))
```
Comment (Xuanwo) [inline metasrv/tests/it/grpc/metasrv_grpc_api.rs:405]:
```suggestion
            .wait(Some(Duration::from_secs(30)))
```

File: metasrv/tests/it/grpc/metasrv_grpc_api.rs [hunk 3]
```diff
@@ -412,6 +397,17 @@ async fn test_auto_sync_addr() -> anyhow::Result<()> {
...
+            .metrics(|m| m.current_leader.is_some(), "a leader is observed")
+            .await?;
+
+        tracing::debug!("got leader, metrics: {:?}", metrics);
```
Comment (Xuanwo) [inline metasrv/tests/it/grpc/metasrv_grpc_api.rs:409]:
```suggestion
        tracing::debug!("got leader, metrics: {metrics:?}");
```

**Review by Xuanwo — APPROVED** [review]
Most LGTM with some optional style-related suggestions.

File: metasrv/tests/it/grpc/metasrv_grpc_api.rs (hunk 2 above)
Comment (Xuanwo) [reply metasrv/tests/it/grpc/metasrv_grpc_api.rs]: BTW, it's quite annoying to wait/sleep in testing code. Is it possible for `raft` to implement something like `notify` or `watch` (test only)?

**Comment by drmingdrmer** [comment]
> BTW, it's quite annoying to wait/sleep in testing code. Is it possible for `raft` to implement something like `notify` or `watch` (test only)? @drmingdrmer

@Xuanwo 
I guess `raft_node.wait(Some(Duration::from_secs(10))).metrics(|x| condition_satisfied(x))` is what you want. It will return once the condition defined in `metrics()` is satisfied or the timeout expired.

The other `sleep()` is to wait for the background task to update the endpoints list.
There are no corresponding metrics for this state. Thus it has to sleep.

Indeed it will be better to use a sleep-check for-loop here at line 413 @ariesdevil :
```rust
        tokio::time::sleep(Duration::from_secs(15)).await;
        let res = client.get_endpoints().await?;
        let res: HashSet<String> = HashSet::from_iter(res.into_iter());
        assert!(
            res.contains(&addr3),
            "endpoints should contains new addr when add node"
        );
```

**Comment by Xuanwo** [comment]
> I guess `raft_node.wait(Some(Duration::from_secs(10))).metrics(|x| condition_satisfied(x))` is what you want. It will return once the condition defined in `metrics()` is satisfied or the timeout expired.

Thanks for the explanation!

**Review by drmingdrmer — APPROVED** [review]

**Comment by Xuanwo** [comment]
Waiting for [test_stateless_standalone_linux](https://github.com/datafuselabs/databend/runs/7073593825?check_suite_focus=true#logs)

**Pull request merged**